File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.js
/**
 * @param {number[]} prices
 * @return {number[]}
 */
var finalPrices = function (prices) {
    const n = prices.length;
    const stack = [];
    const ans = [];

    for (let i = n - 1; i >= 0; i--) {
        while (stack.length && stack[stack.length - 1] > prices[i]) {
            stack.pop();
        }
        ans[i] = stack.length === 0 ? prices[i] : prices[i] - stack[stack.length - 1];
        stack.push(prices[i]);
    }

    return ans;
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        n = len(prices)
        ans = []
        it = 1

        for i in range(n - 1):
            for j in range(i + 1, n):
                if prices[i] >= prices[j]:
                    dif = prices[i] - prices[j]
                    ans.append(dif)
                    break
                if j == n - 1 and prices[i] < prices[j]:
                    ans.append(prices[i])

        ans.append(prices[-1])

        return ans
